File: backend/models/order.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime
from bson import ObjectId
from .user import PyObjectId, Address
from .cart import CartItem

class OrderItem(BaseModel):
    productId: PyObjectId
    name: str
    price: float
    image: str
    quantity: int
    selectedSize: Optional[str] = None
    selectedColor: Optional[str] = None

class OrderCreate(BaseModel):
    items: List[OrderItem]
    deliveryAddress: Address
    paymentMethod: str = "COD"

class Order(BaseModel):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")
    orderId: str = Field(default_factory=lambda: f"MO{str(ObjectId())[-9:]}")
    user: PyObjectId
    items: List[OrderItem]
    totalAmount: float
    status: str = "Confirmed"  # Confirmed, Shipped, In Transit, Delivered, Cancelled
    deliveryAddress: Address
    paymentMethod: str = "COD"
    trackingId: Optional[str] = None
    estimatedDelivery: Optional[datetime] = None
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)
    
    class Config:
        allow_population_by_field_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}

class OrderResponse(BaseModel):
    id: str
    orderId: str
    items: List[OrderItem]
    totalAmount: float
    status: str
    deliveryAddress: Address
    paymentMethod: str
    date: str
    estimatedDelivery: Optional[str] = None
    
    def __init__(self, **data):
        super().__init__(**data)
        if isinstance(data.get('createdAt'), datetime):
            self.date = data['createdAt'].strftime("%Y-%m-%d")
        if data.get('estimatedDelivery') and isinstance(data['estimatedDelivery'], datetime):
            self.estimatedDelivery = data['estimatedDelivery'].strftime("%Y-%m-%d")

File: backend-node/models/Order.js
const mongoose = require('mongoose');

const OrderItemSchema = new mongoose.Schema({
  product: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Product',
    required: true
  },
  name: {
    type: String,
    required: true
  },
  price: {
    type: Number,
    required: true,
    min: [0, 'Price cannot be negative']
  },
  image: {
    type: String,
    required: true
  },
  quantity: {
    type: Number,
    required: true,
    min: [1, 'Quantity must be at least 1']
  },
  selectedSize: {
    type: String,
    default: null
  },
  selectedColor: {
    type: String,
    default: null
  }
});

const AddressSchema = new mongoose.Schema({
  type: {
    type: String,
    required: true
  },
  address: {
    type: String,
    required: true
  },
  city: {
    type: String,
    default: ''
  },
  state: {
    type: String,
    default: ''
  },
  pincode: {
    type: String,
    default: ''
  }
});

const OrderSchema = new mongoose.Schema({
  orderId: {
    type: String,
    unique: true,
    required: true
  },
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  items: [OrderItemSchema],
  totalAmount: {
    type: Number,
    required: true,
    min: [0, 'Total amount cannot be negative']
  },
  status: {
    type: String,
    enum: ['Confirmed', 'Processing', 'Shipped', 'In Transit', 'Delivered', 'Cancelled'],
    default: 'Confirmed'
  },
  deliveryAddress: {
    type: AddressSchema,
    required: true
  },
  paymentMethod: {
    type: String,
    enum: ['COD', 'UPI', 'Card', 'Net Banking'],
    default: 'COD'
  },
  paymentStatus: {
    type: String,
    enum: ['Pending', 'Paid', 'Failed', 'Refunded'],
    default: 'Pending'
  },
  trackingId: {
    type: String,
    default: null
  },
  estimatedDelivery: {
    type: Date,
    default: function() {
      const deliveryDate = new Date();
      deliveryDate.setDate(deliveryDate.getDate() + 7); // 7 days from now
      return deliveryDate;
    }
  },
  deliveredAt: {
    type: Date,
    default: null
  }
}, {
  timestamps: true
});

// Generate order ID before saving
OrderSchema.pre('save', function(next) {
  if (!this.orderId) {
    this.orderId = 'MO' + Date.now() + Math.floor(Math.random() * 1000);
  }
  next();
});

// Index for efficient queries
OrderSchema.index({ user: 1, createdAt: -1 });
OrderSchema.index({ orderId: 1 });
OrderSchema.index({ status: 1 });

module.exports = mongoose.model('Order', OrderSchema);